File: tests/kotlin/test_2_1_1_graph_traversal_iterators.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.apache.tinkerpop.gremlin.structure.Direction
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 2.1.1: Graph Traversal Iterators.
 *
 * Validates vertex and edge iteration, label filtering, and property-based
 * filtering using Kotlin sequences over TinkerCat's native iterators.
 * Mirrors tests/python/test_2_1_1_graph_traversal_iterators.py.
 */
class Test_2_1_1_GraphTraversalIterators : StringSpec({

    lateinit var graph: TinkerCat

    beforeTest { graph = TinkerCat.open() }

    "empty graph has no vertices" {
        graph.vertices().asSequence().toList() shouldHaveSize 0
    }

    "empty graph has no edges" {
        graph.edges().asSequence().toList() shouldHaveSize 0
    }

    "added vertex appears in vertex iterator" {
        val v = graph.addVertex(TinkerVertex.T.label, "person", "name", "Alice")
        val all = graph.vertices().asSequence().toList()
        all shouldHaveSize 1
        all[0].id() shouldBe v.id()
    }

    "vertex count matches actual vertices" {
        repeat(5) { graph.addVertex(TinkerVertex.T.label, "node") }
        graph.vertices.size shouldBe 5
    }

    "edge count matches actual edges" {
        val a = graph.addVertex(TinkerVertex.T.label, "n")
        val b = graph.addVertex(TinkerVertex.T.label, "n")
        val c = graph.addVertex(TinkerVertex.T.label, "n")
        a.addEdge("link", b)
        a.addEdge("link", c)
        graph.edges.size shouldBe 2
    }

    "vertex label is preserved after creation" {
        val v = graph.addVertex(TinkerVertex.T.label, "person", "name", "Bob")
        v.label() shouldBe "person"
    }

    "vertex property is retrievable via value()" {
        val v = graph.addVertex(TinkerVertex.T.label, "person", "name", "Carol", "age", 30)
        v.value<String>("name") shouldBe "Carol"
        v.value<Int>("age") shouldBe 30
    }

    "filter vertices by label using sequence" {
        repeat(3) { graph.addVertex(TinkerVertex.T.label, "person") }
        repeat(2) { graph.addVertex(TinkerVertex.T.label, "software") }
        val persons = graph.vertices().asSequence().filter { it.label() == "person" }.toList()
        persons shouldHaveSize 3
    }

    "filter vertices by property value using sequence" {
        graph.addVertex(TinkerVertex.T.label, "person", "name", "Alice", "age", 25)
        graph.addVertex(TinkerVertex.T.label, "person", "name", "Bob", "age", 30)
        graph.addVertex(TinkerVertex.T.label, "person", "name", "Carol", "age", 35)
        val adults = graph.vertices().asSequence()
            .filter { it.value<Int>("age") != null && it.value<Int>("age")!! >= 30 }
            .toList()
        adults shouldHaveSize 2
    }

    "edge connects correct out and in vertices" {
        val a = graph.addVertex(TinkerVertex.T.label, "person", "name", "Alice")
        val b = graph.addVertex(TinkerVertex.T.label, "person", "name", "Bob")
        a.addEdge("knows", b)
        val edge = graph.edges().asSequence().first()
        edge.outVertex().id() shouldBe a.id()
        edge.inVertex().id() shouldBe b.id()
    }

    "edge label is preserved" {
        val a = graph.addVertex(TinkerVertex.T.label, "n")
        val b = graph.addVertex(TinkerVertex.T.label, "n")
        a.addEdge("knows", b)
        graph.edges().asSequence().first().label() shouldBe "knows"
    }

    "vertex out-edges are navigable" {
        val a = graph.addVertex(TinkerVertex.T.label, "person")
        val b = graph.addVertex(TinkerVertex.T.label, "person")
        val c = graph.addVertex(TinkerVertex.T.label, "person")
        a.addEdge("knows", b)
        a.addEdge("knows", c)
        val outEdges = a.edges(Direction.OUT).asSequence().toList()
        outEdges shouldHaveSize 2
    }

    "vertex in-edges are navigable" {
        val a = graph.addVertex(TinkerVertex.T.label, "person")
        val b = graph.addVertex(TinkerVertex.T.label, "person")
        b.addEdge("knows", a)
        val inEdges = a.edges(Direction.IN).asSequence().toList()
        inEdges shouldHaveSize 1
    }

    "missing property returns null" {
        val v = graph.addVertex(TinkerVertex.T.label, "node")
        v.value<String>("nonexistent") shouldBe null
    }

    "vertex id is non-null" {
        val v = graph.addVertex(TinkerVertex.T.label, "node")
        v.id() shouldNotBe null
    }

    "multiple vertices have distinct ids" {
        val v1 = graph.addVertex(TinkerVertex.T.label, "node")
        val v2 = graph.addVertex(TinkerVertex.T.label, "node")
        v1.id() shouldNotBe v2.id()
    }
})

File: tests/test_2_1_1_graph_traversal_iterators.py
"""
Tests for Task 2.1.1: Graph Traversal Iterators.

Verifies lazy iterator behaviour, direction/label filtering, and index-backed
property lookups.  See docs/project/changelog/task-2.1.1-graph-traversal-iterators.adoc.
"""

import pytest

from tinkercat import TinkerCat

from mocks import build_modern_graph

@pytest.fixture
def modern():
    with TinkerCat() as g:
        build_modern_graph(g)
        yield g

def test_vertices_returns_all_vertices(modern):
    assert sum(1 for _ in modern.vertices()) == 6

def test_edges_returns_all_edges(modern):
    assert sum(1 for _ in modern.edges()) == 6

def test_vertex_filter_by_label(modern):
    people = [v for v in modern.vertices() if v.label == "person"]
    assert len(people) == 4

def test_vertex_filter_by_property(modern):
    old = [v for v in modern.vertices() if (v.value("age") or 0) > 30]
    assert len(old) == 2  # josh(32), peter(35)

def test_out_edges_direction_filtering(modern):
    marko = next(v for v in modern.vertices() if v.value("name") == "marko")
    out = list(marko.out_edges())
    assert len(out) == 3

def test_out_edges_label_filtering(modern):
    marko = next(v for v in modern.vertices() if v.value("name") == "marko")
    knows = list(marko.out_edges("knows"))
    assert len(knows) == 2

def test_in_edges_direction_filtering(modern):
    lop = next(v for v in modern.vertices() if v.value("name") == "lop")
    in_edges = list(lop.in_edges())
    assert len(in_edges) == 3

def test_out_vertices_navigation(modern):
    marko = next(v for v in modern.vertices() if v.value("name") == "marko")
    neighbors = list(marko.out_vertices("knows"))
    names = {v.value("name") for v in neighbors}
    assert names == {"vadas", "josh"}

def test_iterator_is_reusable_from_graph(modern):
    first  = sum(1 for _ in modern.vertices())
    second = sum(1 for _ in modern.vertices())
    assert first == second

def test_property_value_retrieval(modern):
    marko = next(v for v in modern.vertices() if v.value("name") == "marko")
    assert marko.value("age") == 29

def test_edge_property_retrieval(modern):
    marko = next(v for v in modern.vertices() if v.value("name") == "marko")
    knows_edges = list(marko.out_edges("knows"))
    weights = {e.value("weight") for e in knows_edges}
    assert weights == {0.5, 1.0}
